components/application-tag-mgt/org.wso2.carbon.identity.application.tag.mgt/src/main/java/org/wso2/carbon/identity/application/tag/mgt/dao/impl/ApplicationTagDAOImpl.java

```diff
@@ -0,0 +1,256 @@
+        if (log.isDebugEnabled()) {
+            log.debug("Creating Application Tag of " + applicationTagDTO.getName());
+        }
+        String generatedAppTagId = UUID.randomUUID().toString();
+        try (Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true)) {
+            try {
+                PreparedStatement prepStmt = dbConnection.prepareStatement(SQLConstants.ADD_APPLICATION_TAG);
+                prepStmt.setString(1, generatedAppTagId);
+                prepStmt.setString(2, applicationTagDTO.getName());
+                prepStmt.setString(3, applicationTagDTO.getColour());
+                prepStmt.setInt(4, tenantId);
+                prepStmt.execute();
+
+                ResultSet results = prepStmt.getGeneratedKeys();
+
+                String applicationTagId = null;
+                if (results.next()) {
+                    applicationTagId = results.getString(1);
+                }
+                // some JDBC Drivers returns this in the result, some don't
+                if (applicationTagId == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("JDBC Driver did not return the tag id, executing Select operation");
+                    }
+                    applicationTagId = getApplicationTagByName(applicationTagDTO.getName(), tenantId, dbConnection);
+                }
+
+                IdentityDatabaseUtil.commitTransaction(dbConnection);
+                return applicationTagId;
+            } catch (SQLException e) {
+                IdentityDatabaseUtil.rollbackTransaction(dbConnection);
+                if (e.getMessage().toLowerCase().contains(SQLConstants.APP_TAG_UNIQUE_CONSTRAINT.toLowerCase())) {
+                    throw ApplicationTagManagementUtil.handleClientException(
+                            ApplicationTagManagementConstants.ErrorMessages.ERROR_CODE_APP_TAG_ALREADY_EXISTS,
+                            String.valueOf(tenantId));
+                }
+                // Handle the DB2 database unique constraint violation error.
+                if (dbConnection.getMetaData().getDatabaseProductName().contains("DB2")) {
+                    if (e.getMessage().contains(SQLConstants.DB2_SQL_ERROR_CODE_UNIQUE_CONSTRAINT)) {
+                        throw ApplicationTagManagementUtil.handleClientException(
+                                ApplicationTagManagementConstants.ErrorMessages.ERROR_CODE_APP_TAG_ALREADY_EXISTS,
+                                String.valueOf(tenantId));
+                    }
+                }
+                throw e;
+            }
+        } catch (SQLException e) {
+            throw ApplicationTagManagementUtil.handleServerException(
+                    ApplicationTagManagementConstants.ErrorMessages.ERROR_CODE_ERROR_WHILE_ADDING_APP_TAG, e);
+        }
+    }
+
+    @Override
+    public List<ApplicationTagsListItem> getAllApplicationTags(String tenantDomain) throws ApplicationTagMgtException {
```

Comment:
Implementing a method to return all the resources of a collection can be bad.

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/4904/commits/f4c723bc4dd30679a10beca35a8da023f4184aed
Introduced a ApplicationTagListResponse